### PR TITLE
BUGFIX: Fix the NodeTypes schema files and allow properties `search` and `options`

### PR DESCRIPTION
**Pull request by mficzel**
The NodeType schemas can now handle the case that a property in a NodeType definition can be either a dictionary or null.

Additionally the keys `search`  and `options` as extension point for package-configuration.

The key `search` is used by several search implementations and currently causes validation errors.
The key `options` is allowed to avoid such problems in the future and to create a place where packages can place custom configuration.

### CODE REVIEW

File: TYPO3.TYPO3CR/Resources/Private/Schema/NodeTypes.schema.yaml [hunk 1]
```diff
@@ -65,6 +65,16 @@ additionalProperties:
             # we intentionally do not specify which properties are allowed here;
             # as other packages (such as TYPO3.Neos) might do that instead.
 
+          'options':
+            type: dictionary
+            # we intentionally do not specify which properties are allowed here;
+            # as other packages might do that instead.
+
+          'search':
```
Comment (aertmann) [inline TYPO3.TYPO3CR/Resources/Private/Schema/NodeTypes.schema.yaml]: this doesn't belong in the core as it's not used by the core.. instead it should be added in the `content-repository-search` package

File: TYPO3.TYPO3CR/Resources/Private/Schema/NodeTypes.schema.yaml [hunk 2]
```diff
@@ -65,6 +65,16 @@ additionalProperties:
             # we intentionally do not specify which properties are allowed here;
             # as other packages (such as TYPO3.Neos) might do that instead.
 
+          'options':
+            type: dictionary
+            # we intentionally do not specify which properties are allowed here;
+            # as other packages might do that instead.
```
Comment (aertmann) [inline TYPO3.TYPO3CR/Resources/Private/Schema/NodeTypes.schema.yaml]: this is not used in the core AFAIK, thus doesn't really belong here.. but you could argue it make sense as a wildcard

File: TYPO3.TYPO3CR/Resources/Private/Schema/NodeTypes.schema.yaml [hunk 3]
```diff
@@ -74,6 +84,13 @@ additionalProperties:
 
     'options':
       type: dictionary
+      # we intentionally do not specify which properties are allowed here;
+      # as other packages might do that instead.
+
+    'search':
+      type: dictionary
+      # we intentionally do not specify which properties are allowed here;
+      # as other packages might do that instead.
```
Comment (aertmann) [inline TYPO3.TYPO3CR/Resources/Private/Schema/NodeTypes.schema.yaml]: same here

**Comment by mficzel** [comment]
@aertmann each schema is validated seperately and since we use `additionalProperties: FALSE` we have to allow each key explicitly. In cases where we know that packages will define details we define the main-key as dictionary with a note that specifics are defined elsewhere. That is what i did for the additional keys in this pr.

The `search` key is already used by the *Search-Packages. If you are installing any Search Package you immediately get Errors in NodeType-Configuration-Validation because this key `search` is not allowed for properties. I think this can be considered buggish and is the reason why i based the pr on 2.2.

An `options` key is already used in several other places even in nodeType-Configuration. My reason for allowing this for property-configurations is to have a defined place where packages can place custom configurations without causing validation trouble. I added this one to the schema to allow Packages to use extra configuration without breaking the main validation.

**Comment by mficzel** [comment]
BTW: I added the needs feedback label because i am not sure that 2.2 is the correct target branch.

**Comment by aertmann** [comment]
Okay I see that they are validated individually, not sure what to do about that tbh.

It might be better to set `additionalAttributes` to `true` instead and just ensure that those keys that do exist are valid instead of dying unknown ones. E.g. I only get an issue when adding search to the root of a node type, not options or search to any other places.

**Comment by mficzel** [comment]
@aertmann i wonder why search on properties does not cause a validation error because looking at the schema it should. I will check the schema again.

The idea of not setting `additionalProperties = true` is to be able to provide errors and help integrators if properties are placed wrong or are misspelled. That is the main idea of the whole schema thing. That is why i think that adding a generic playground section `options` wich packages can use and provide theirs own subschema is a good idea.

**Comment by mficzel** [comment]
@aertmann as expected the schema was broken. I adjusted the pr to fix the schema and alongside allow search and options. 

To make the review easier here an examples for what was wrong with the schema.  

Wrong schema:
```
type: ['null', 'dictionary']
additionalProperties: FALSE
properties: ...
```
Correct schema:
```
type:
 - 
   type: 'null'
 - 
    type: 'dictionary'
    additionalProperties: FALSE
    properties: ...
```

**Comment by aertmann** [comment]
looks good and works as expected.. need to retarget for 2.3 though

**Comment by mficzel** [comment]
I changed the target branch to 2.3 since 2.2 is out of maintenance now.

**Pull request merged**